src/Symfony/Component/Security/Core/Authentication/ContainerAwareAuthenticationProviderManager.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authentication;
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Uses the Container to lazily instantiate the providers.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ContainerAwareAuthenticationProviderManager extends AbstractAuthenticationProviderManager
+{
+    private $providerServiceIds;
+
+    private $container;
+
+    private $eraseCredentials;
+
+    public function __construct(array $providerServiceIds, ContainerInterface $container, $eraseCredentials = true)
+    {
+        $this->providerServiceIds = $providerServiceIds;
+        $this->container = $container;
+        $this->eraseCredentials = $eraseCredentials;
+    }
+
+    protected function getProviders()
+    {
+        $providers = array();
+        foreach ($this->providerServiceIds as $serviceId) {
+            $providers[] = $this->container->get($serviceId);
+        }
```

Comment:
Yes I do, but I mean more of the internal working. My understanding of yield:
 1. Goes through the generator method until it finds a yield
 2. Then goes into the foreach until that iteration is done
 3. Goes back into the generator method until it finds a yield
 4. goto 2

This is how I assume it works because it also reduces memory usage by not having to build an array, hence I believe it would be a benefit to use here (for both memory and method calls) as it's always used in an iteration.

Downside, `if (!$providers) {` can't be used. The only way to determine if there's no providers, is to see if the foreach was triggered or not.